haystack/components/tools/tool_invoker.py

```diff
@@ -480,83 +557,93 @@ def run(
-                    continue
-
-                tool_to_invoke = self._tools_with_names[tool_name]
-
-                # 1) Combine user + state inputs
-                llm_args = tool_call.arguments.copy()
-                final_args = self._inject_state_args(tool_to_invoke, llm_args, state)
-
-                # Check whether to inject streaming_callback
-                if (
-                    resolved_enable_streaming_passthrough
-                    and streaming_callback is not None
-                    and "streaming_callback" not in final_args
-                ):
-                    invoke_params = self._get_func_params(tool_to_invoke)
-                    if "streaming_callback" in invoke_params:
-                        final_args["streaming_callback"] = streaming_callback
-
-                # 2) Invoke the tool
-                try:
-                    tool_result = tool_to_invoke.invoke(**final_args)
-
-                except ToolInvocationError as e:
-                    error_message = self._handle_error(e)
-                    tool_messages.append(ChatMessage.from_tool(tool_result=error_message, origin=tool_call, error=True))
-                    continue
-
-                # 3) Merge outputs into state
-                try:
-                    self._merge_tool_outputs(tool_to_invoke, tool_result, state)
-                except Exception as e:
-                    try:
-                        error_message = self._handle_error(
-                            ToolOutputMergeError(f"Failed to merge tool outputs from tool {tool_name} into State: {e}")
-                        )
+        # Collect all tool calls and their parameters for parallel execution
+        tool_call_params, error_messages = self._prepare_tool_call_params(
+            messages_with_tool_calls, state, streaming_callback, resolved_enable_streaming_passthrough
+        )
+        tool_messages.extend(error_messages)
+
+        # 3) Execute valid tool calls in parallel
+        if tool_call_params:
+            with ThreadPoolExecutor(max_workers=self.max_workers) as executor:
+                futures = []
+                for params in tool_call_params:
+                    future = executor.submit(self._execute_single_tool_call, **params)  # type: ignore[arg-type]
+                    futures.append(future)
+
+                # Process results as they complete
+                for future in futures:
+                    result = future.result()
+                    if isinstance(result, ChatMessage):
+                        tool_messages.append(result)
+                    else:
+                        # Handle state merging and prepare tool result message
+                        tool_call, tool_to_invoke, tool_result = result
+
+                        # 4) Merge outputs into state
+                        with self._state_lock:
+                            try:
+                                self._merge_tool_outputs(tool_to_invoke, tool_result, state)
```

Comment:
I'm a little confused. Since we are looping through futures and from what I understand calling `future.result()` blocks sequentially until each future completes, I'd be under the impression that we don't need the threading lock here right?

cc @mpangrazzi